contextualized/regression/datasets.py

```diff
@@ -1,121 +1,187 @@
+        super().__init__(C, X, Y, idx=idx)
+        self._generate_samples()
 
-class MultitaskMultivariateDataset(Dataset):
-    """
-    Multi-task Multivariate Dataset.
-    """
+    def _generate_samples(self):
+        self.n_i = 0
+        self.x_i = 0
+        self.y_i = 0
+        self.all_samples = []
 
-    def __next__(self):
+        for i in range(self._total_len()):
+            self.all_samples.append(self._get_next_item(i))
+
+        # choose subset based on n_i matching idx
+        self.samples = [s for s in self.all_samples if s[1] in self.idx]
+
+    def _get_next_item(self, i):
         if self.y_i >= self.y_dim:
             self.n_i += 1
             self.y_i = 0
         if self.n_i >= self.n:
             self.n_i = 0
-            raise StopIteration
         t = torch.zeros(self.y_dim)
         t[self.y_i] = 1
+
+        ret = (t, self.n_i, self.y_i)
+        self.y_i += 1
+        return ret
+
+    def _total_len(self):
+        return self.n * self.y_dim
+
+    def __len__(self):
+        return len(self.samples)
+
+    def __getitem__(self, idx):
+        t, n_i, y_i = self.samples[idx]
         ret = (
-            self.C[self.n_i],
+            self.C[n_i],
             t,
-            self.X[self.n_i],
-            self.Y[self.n_i, self.y_i].unsqueeze(0),
-            self.n_i,
-            self.y_i,
+            self.X[n_i],
+            self.Y[n_i, y_i].unsqueeze(0),
+            n_i,
+            y_i,
         )
-        self.y_i += 1
         return ret
 
+
+class UnivariateDataset(RegressionDatasetBase):
+    def __init__(self, C, X, Y, idx=[], dtype=torch.float):
+        super().__init__(C, X, Y, idx=idx)
+        self._generate_samples()
+
+    def _generate_samples(self):
+        self.n_i = 0
+        self.x_i = 0
+        self.y_i = 0
+        self.all_samples = []
+
+        for i in range(self._total_len()):
+            self.all_samples.append(self._get_next_item(i))
+
+        self.samples = [self.all_samples[i] for i in self.idx]
+
+    def _get_next_item(self, i):
+        if self.n_i >= self.n:
+            self.n_i = 0
+        ret = self.n_i
+        self.n_i += 1
+        return ret
+
+    def _total_len(self):
+        return self.n
+
     def __len__(self):
-        return self.n * self.y_dim
+        return len(self.samples)
 
+    def __getitem__(self, idx):
+        n_i = self.samples[idx]
+        ret = (
+            self.C[n_i],
+            self.X[n_i].expand(self.y_dim, -1).unsqueeze(-1),
+            self.Y[n_i].expand(self.x_dim, -1).T.unsqueeze(-1),
+            n_i,
+        )
+        return ret
 
-class MultitaskUnivariateDataset(Dataset):
-    """
-    Multitask Univariate Dataset
-    """
 
-    def __next__(self):
+class MultitaskUnivariateDataset(RegressionDatasetBase):
+    def __init__(self, C, X, Y, idx=[], dtype=torch.float):
+        super().__init__(C, X, Y, idx)
+        self._generate_samples()
+
+    def _generate_samples(self):
```

Comment:
Move the code from `_generate_samples` into `__init__`. There's not enough code duplication in `__init__` to justify this being a separate method. 

FYI: If we did use `_generate_samples`, we would also want to remove `__init__` from each of the subclasses and put a generic `_generate_samples` in the base class. We could call this from the base class `__init__` and re-implement `_generate_samples` in each subclass. But again, this only makes sense if we would have a lot of code duplication re-writing `__init__` in each of the subclasses.